app/controllers/holdings_controller.rb

```diff
@@ -0,0 +1,103 @@
+class HoldingsController < ApplicationController
+    before_action :authenticate_user!
+
+    def new
+        @stock = Stock.find(params[:stock_id])
+        @holding = current_user.holdings.build
+        @wallet = current_user.wallet
+        @buy_value = helpers.stock_total_value(true, @stock.current_price)
+        @sell_value = helpers.stock_total_value(false, @stock.current_price)
+        @current_stock = current_user.holdings.find_by(ticker: @stock.ticker)
+
+    end
+
+    def create 
+        currently_holding_stock = current_user.holdings.find_by(ticker: params[:holding][:ticker])
+        trade_logic(currently_holding_stock)
+    end
+
+    private
+
+    def holding_params
+        params.require(:holding).permit(:user_id, :ticker, :log_price, :amount, :stock_id, :transaction_type, :units)
+    end
+
+    def convert_to_units(hold, stock_price) 
+        hold / stock_price
+    end
+
+    def trade_logic(currently_holding_stock)
+        @user_wallet = current_user.wallet
+        case params[:holding][:transaction_type]
+        when 'BUY'
+          buy_logic(currently_holding_stock)
+        when 'SELL'
+          sell_logic(currently_holding_stock)
+        end
+    end
+
+    def buy_logic(currently_holding_stock) 
+        if currently_holding_stock.nil?
+            @holding = current_user.holdings.build(holding_params)
+            @holding.units = @holding.amount / @holding.log_price
+            @user_wallet.balance = current_user.wallet.balance - @holding.amount
+            @holding.save if @user_wallet.balance.positive?
+            buy_logic_update(@user_wallet, @holding, @holding.units)
+          else
+            @holding = currently_holding_stock
+            amount = params[:holding][:amount].to_f
+            stock_price = params[:holding][:log_price].to_f
+            units_to_be_added = amount / stock_price
+            @holding.units = (@holding.units + units_to_be_added)
+            @holding.amount = (@holding.amount + params[:holding][:amount].to_f)
+            @user_wallet.balance = current_user.wallet.balance - params[:holding][:amount].to_f
+            @holding.save if @user_wallet.balance.positive?
+            buy_logic_create(@user_wallet, @holding, units_to_be_added)
+        end
+    end
+
```

Comment:
currently_holding_stock => checks if the user already has similar stocks in their holdings. 
If they don't we will create new holding entry for the stock (as long has the can afford it)
ELSE we will update the existing stock holding entry in their account.

**params[:holding][:columns] => takes the entered values in the trading form
![image](https://user-images.githubusercontent.com/86232999/148647157-60ff37b2-4fd5-4811-93e8-3b3f7b0fbcd7.png)